runtime/builder/golang/golang.go

```diff
@@ -0,0 +1,189 @@
+package golang
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/micro/go-micro/v3/runtime/builder"
+	"github.com/micro/go-micro/v3/runtime/local"
+)
+
+// NewBuilder returns a golang builder which can build a go binary given some source
+func NewBuilder() (builder.Builder, error) {
+	path, err := locateGo()
+	if err != nil {
+		return nil, fmt.Errorf("Error locating go binary: %v", err)
+	}
+
+	return &golang{
+		cmdPath: path,
+		tmpDir:  os.TempDir(),
+	}, nil
+}
+
+type golang struct {
+	cmdPath string
+	tmpDir  string
+}
+
+// Build a binary using source. If an archive was used, e.g. tar, this should be specified in the
+// options. If no archive option is passed, the builder will treat the source as a single file.
+func (g *golang) Build(src io.Reader, opts ...builder.Option) (io.Reader, error) {
+	// parse the options
+	var options builder.Options
+	for _, o := range opts {
+		o(&options)
+	}
+
+	// create a tmp dir to contain the source
+	dir, err := ioutil.TempDir(g.tmpDir, "src")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(dir)
+
+	// decode the source and write to the tmp directory
+	switch options.Archive {
+	case "":
+		err = writeFile(src, dir)
+	case "tar":
+		err = unarchiveTar(src, dir)
+	case "zip":
+		err = unarchiveZip(src, dir)
+	default:
+		return nil, errors.New("Invalid Archive")
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// determine the entrypoint if one wasn't provided
+	if len(options.Entrypoint) == 0 {
+		ep, err := local.Entrypoint(dir)
+		if err != nil {
+			return nil, err
+		}
+		options.Entrypoint = ep
+	}
+
+	// build the binary
+	cmd := exec.Command(g.cmdPath, "build", "-o", "build", filepath.Dir(options.Entrypoint))
```

Comment:
If the archive that we decompress has a directory called build I think this might fail. So I think we might want to pick something with a name that's less likely to clash. `micro_build` or something

```
The -o flag forces build to write the resulting executable or object
to the named output file or directory, instead of the default behavior described
in the last two paragraphs. If the named output is a directory that exists,
then any resulting executables will be written to that directory.
```